picoquic/frames.c

```diff
@@ -1888,211 +1896,140 @@ int picoquic_prepare_path_challenge_frame(uint8_t* bytes,
+            /* TODO: verify that the network addresses match the path */
+            found_challenge = 1;
+            cnx->path[i]->challenge_verified = 1;
         }
     }
 
-    *consumed = byte_index;
+    if (!found_challenge && cnx->callback_fn != NULL) {
+        cnx->callback_fn(cnx, 0, (void *)&response, sizeof(response), picoquic_callback_challenge_response, cnx->callback_ctx);
+    }
 
-    return ret;
+    return 0;
 }
 
+
+#define FRAME_OPS(name, ack_needed, restricted_ok) \
+    {picoquic_decode_##name##_frame, ack_needed, restricted_ok}
+#define NO_ACK 0
+#define DO_ACK 1
+#define NO_RESTRICTED 0
+#define OK_RESTRICTED 1
+#define DUP16(v) v, v, v, v, v, v, v, v, v, v, v, v, v, v, v, v
+
+static const struct frame_ops_t {
+  picoquic_decode_frame_t decode;
+  uint8_t                 ack_needed;
+  uint8_t                 restricted_ok;
+} frame_ops[256] = {
+  FRAME_OPS(padding,           NO_ACK, OK_RESTRICTED),
+  FRAME_OPS(reset_stream,      DO_ACK, NO_RESTRICTED),
+  FRAME_OPS(connection_close,  NO_ACK, OK_RESTRICTED),
+  FRAME_OPS(application_close, NO_ACK, NO_RESTRICTED),
+  FRAME_OPS(max_data,          DO_ACK, NO_RESTRICTED),
+  FRAME_OPS(max_stream_data,   DO_ACK, NO_RESTRICTED),
+  FRAME_OPS(max_stream_id,     DO_ACK, NO_RESTRICTED),
+  FRAME_OPS(ping,              DO_ACK, NO_RESTRICTED),  /* Not ok for restricted? */
+  FRAME_OPS(blocked,           DO_ACK, NO_RESTRICTED),
+  FRAME_OPS(stream_blocked,    DO_ACK, NO_RESTRICTED),
+  FRAME_OPS(stream_id_needed,  DO_ACK, NO_RESTRICTED),
+  FRAME_OPS(new_connection_id, DO_ACK, NO_RESTRICTED),
+  FRAME_OPS(stop_sending,      DO_ACK, NO_RESTRICTED),
+  FRAME_OPS(ack,               NO_ACK, OK_RESTRICTED),
+  FRAME_OPS(path_challenge,    NO_ACK, OK_RESTRICTED),
+  FRAME_OPS(path_response,     NO_ACK, OK_RESTRICTED),
+  DUP16(FRAME_OPS(stream,      DO_ACK, OK_RESTRICTED)),  /* 0x10 - 0x1f */
+  DUP16({0}),                                            /* 0x20 - 0x2f */
+  DUP16({0}),                                            /* 0x30 - 0x3f */
+  DUP16({0}),                                            /* 0x40 - 0x4f */
+  DUP16({0}),                                            /* 0x50 - 0x5f */
+  DUP16({0}),                                            /* 0x60 - 0x6f */
+  DUP16({0}),                                            /* 0x70 - 0x7f */
+  DUP16({0}),                                            /* 0x80 - 0x8f */
+  DUP16({0}),                                            /* 0x90 - 0x9f */
+  DUP16({0}),                                            /* 0xa0 - 0xaf */
+  DUP16({0}),                                            /* 0xb0 - 0xbf */
+  DUP16({0}),                                            /* 0xc0 - 0xbf */
+  DUP16({0}),                                            /* 0xd0 - 0xdf */
+  DUP16({0}),                                            /* 0xe0 - 0xef */
+  {0}, {0}, {0}, {0}, {0}, {0}, {0}, {0},
+  {0}, {0}, {0}, {0}, {0}, {0}, {0}, {0}
+};
+
```

Comment:
Can I suggest that we do not implement the function table in the current PR? I would rather break this current PR in two:

1) Change the calling logic for some frame types to use the pointer/pointer max template, intead of the byte/length/consumed template, but do not change the signature of the picoquic_decode_frames function. Just keep the same switch statement, calling the new functions instead of the old ones.

2) Once the first PR is verified, change from the switch statement to a function table resolution.

The advantage is that the first changes will be almost entirely internal to "frames.c", with no need to change any of the functions that consumes "picoquic_decode_frames". In particular, there will not be a need to change any of the test cases. We will be able to verify that the decoding works.

In fact, we could consider doing the same thing with the "skip" functions, again before moving to the table structure. That will also allow us to use the commonality between "decode with no side effect" and "skip".

Reply:
This is a reasonable suggestion.

I was going to move the skip functions into the table as well.  Just did not get to them and decided to focus on decode functions first.

Reply:
OK, do what is easiest. The problem is not so much the table as the need to keep the existing external API for now, so we don't have to change the protocol implementation or the tests. As long as we achieve that, I am fine.

Reply:
The "external API" is not super-well defined.  APIs for callbacks is one thing (reasonably clear -- external API).  But one test is calling picoquic_decode_stream_frame directly.  That's a layer violation, so it may need to change.  (I had almost marked all picoquic_decode_*_frame functions static.)